arangod/RestHandler/RestDumpHandler.cpp

```diff
@@ -0,0 +1,269 @@
+  size_t const len = suffixes.size();
+
+  if (type == rest::RequestType::DELETE_REQ) {
+    if (len == 1) {
+      // end a dump
+      handleCommandDumpFinished();
+    } else {
+      generateError(
+          Result(TRI_ERROR_BAD_PARAMETER, "expecting DELETE /_api/dump/<id>"));
+    }
+  } else if (type == rest::RequestType::POST) {
+    if (len == 1 && suffixes[0] == "start") {
+      // start a dump
+      handleCommandDumpStart();
+    } else if (len == 2 && suffixes[0] == "next") {
+      // fetch next data from a dump
+      handleCommandDumpNext();
+    } else {
+      generateError(
+          Result(TRI_ERROR_BAD_PARAMETER,
+                 "expecting POST /_api/dump/start or /_api/dump/next/<id>"));
+    }
+  } else {
+    // invalid HTTP method
+    generateError(rest::ResponseCode::METHOD_NOT_ALLOWED,
+                  TRI_ERROR_HTTP_METHOD_NOT_ALLOWED);
+  }
+
+  return RestStatus::DONE;
+}
+
+RequestLane RestDumpHandler::lane() const {
+  auto type = _request->requestType();
+
+  if (type == rest::RequestType::DELETE_REQ) {
+    // delete should be prioritized, because it frees
+    // up resources
+    return RequestLane::CLUSTER_INTERNAL;
+  }
+  return RequestLane::SERVER_REPLICATION;
+}
+
+/// @brief returns the short id of the server which should handle this request
+ResultT<std::pair<std::string, bool>> RestDumpHandler::forwardingTarget() {
+  auto base = RestVocbaseBaseHandler::forwardingTarget();
+  if (base.ok() && !std::get<0>(base.get()).empty()) {
+    return base;
+  }
+
+  auto type = _request->requestType();
+  auto const& suffixes = _request->suffixes();
+  size_t const len = suffixes.size();
+  if (type == rest::RequestType::DELETE_REQ) {
+    if (len != 1) {
+      return ResultT<std::pair<std::string, bool>>::error(
+          TRI_ERROR_BAD_PARAMETER, "expecting DELETE /_api/dump/<id>");
+    }
+  } else if (type == rest::RequestType::POST) {
+    if ((len == 1 && suffixes[0] != "start") ||
+        (len == 2 && suffixes[0] != "next") || len < 1 || len > 2) {
+      return ResultT<std::pair<std::string, bool>>::error(
+          TRI_ERROR_BAD_PARAMETER,
+          "expecting POST /_api/dump/start or /_api/dump/next/<id>");
+    }
+  } else {
+    // invalid HTTP method
+    return ResultT<std::pair<std::string, bool>>::error(
+        TRI_ERROR_HTTP_METHOD_NOT_ALLOWED);
+  }
+
+  if (ServerState::instance()->isCoordinator()) {
+    ServerID const& DBserver = _request->value("dbserver");
+    if (!DBserver.empty()) {
+      // if DBserver property present, remove auth header
+      return std::make_pair(DBserver, true);
+    }
+
+    return ResultT<std::pair<std::string, bool>>::error(
+        TRI_ERROR_BAD_PARAMETER, "need a 'dbserver' parameter");
+  }
+
+  return {std::make_pair(StaticStrings::Empty, false)};
+}
+
+void RestDumpHandler::handleCommandDumpStart() {
+  std::string database = _request->databaseName();
+  std::string user = _request->user();
+
+  // TODO: check permissions
```

Comment:
I guess this TODO still needs to be addressed.

Reply:
Oh wow, I wasn't aware of that. My plan was to add a permissions test as well. Would have found out later 😆

Reply:
Permissions + Test added.